File: Permutations - Leetcode 46/Permutations - Leetcode 46.js
function permute(nums) {
    const ans = [];
    const sol = [];
    const used = Array(nums.length).fill(false);

    function backtrack() {
        if (sol.length === nums.length) {
            ans.push([...sol]);
            return;
        }

        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            
            sol.push(nums[i]);
            used[i] = true;
            backtrack();
            sol.pop();
            used[i] = false;
        }
    }

    backtrack();
    return ans;
}

File: Permutations - Leetcode 46/Permutations - Leetcode 46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res, sol = [], []

        def backtrack():
            if len(sol) == n:
                res.append(sol[:])
                return

            for x in nums:
                if x not in sol:
                    sol.append(x)
                    backtrack()
                    sol.pop()

        backtrack()
        return res

# Time Complexity: O(n!)
# Space Complexity: O(n)
